tests_integration/test_intro.py

```diff
@@ -0,0 +1,273 @@
+from typing import Any
+
+import dspy
+
+
+class TestIntroIntegration:
+    def test_dspy_workflow(self) -> None:
+        self.setup_dspy()
+
+        dev_example, dev_set, training_set = self.assert_dataset_loading()
+
+        self.assert_basic_qa(dev_example)
+
+        self.assert_retrieval(dev_example)
+
+        self.assert_compilation(dev_set, training_set)
+
+    def assert_compilation(self, devset, trainset) -> None:
+        class GenerateAnswer(dspy.Signature):
+            """Answer questions with short factoid answers."""
+
+            context = dspy.InputField(desc="may contain relevant facts")
+            question = dspy.InputField()
+            answer = dspy.OutputField(desc="often between 1 and 5 words")
+
+        class RAG(dspy.Module):
+            def __init__(self, num_passages=3):
+                super().__init__()
+
+                self.retrieve = dspy.Retrieve(k=num_passages)
+                self.generate_answer = dspy.ChainOfThought(GenerateAnswer)
+
+            def forward(self, question):
+                context = self.retrieve(question).passages
+                prediction = self.generate_answer(context=context, question=question)
+                return dspy.Prediction(context=context, answer=prediction.answer)
+
+        from dspy.teleprompt import BootstrapFewShot
+
+        # Validation logic: check that the predicted answer is correct.
+        # Also check that the retrieved context actually contains that answer.
+        def validate_context_and_answer(example, pred, trace=None):  # noqa
+            answer_em = dspy.evaluate.answer_exact_match(example, pred)
+            answer_pm = dspy.evaluate.answer_passage_match(example, pred)
+            return answer_em and answer_pm
+
+        # Set up a basic teleprompter, which will compile our RAG program.
+        teleprompter = BootstrapFewShot(metric=validate_context_and_answer)
+        # Compile the RAG model
+        compiled_rag = teleprompter.compile(RAG(), trainset=trainset)
+
+        # Test the compiled RAG model with a question
+        my_question = "What castle did David Gregory inherit?"
+        pred = compiled_rag(my_question)
+
+        # Assertions to verify the compiled RAG model
+        assert f"Question: {my_question}" == "Question: What castle did David Gregory inherit?"
+        assert f"Predicted Answer: {pred.answer}" == "Predicted Answer: Kinnairdy Castle"
+        assert (
+            f"Retrieved Contexts (truncated): {[c[:10] + '...' for c in pred.context]}"
+            == "Retrieved Contexts (truncated): ['David Greg...', 'Gregory Ta...', 'David Greg...']"
+        )
+
+        # Verify compiled model's parameters
+        for name, parameter in compiled_rag.named_predictors():
+            assert name is not None
```

Comment:
i think it's good to also check that the `name` and `parameter.demos[0]` match up with the cached objects from the notebook.